include/deal.II/lac/petsc_snes.templates.h

```diff
@@ -0,0 +1,458 @@
+    AMatrixType &A,
+    PMatrixType &P)
+  {
+    this->A = &A;
+    this->P = &P;
+  }
+
+
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  unsigned int
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::solve(VectorType &x)
+  {
+    auto snes_function_ =
+      [](SNES snes, Vec x, Vec f, void *ctx) -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      (void)snes;
+      auto user = static_cast<NonlinearSolver *>(ctx);
+
+      VectorType xdealii(x);
+      VectorType fdealii(f);
+      AssertUser(user->residual(xdealii, fdealii), "residual");
+      petsc_object_state_increase(f);
+      PetscFunctionReturn(0);
+    };
+
+    auto snes_jacobian_ =
+      [](SNES snes, Vec x, Mat A, Mat P, void *ctx) -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      (void)snes;
+      auto user = static_cast<NonlinearSolver *>(ctx);
+
+      VectorType  xdealii(x);
+      AMatrixType Adealii(A);
+      PMatrixType Pdealii(P);
+      AssertUser(user->jacobian(xdealii, Adealii, Pdealii), "jacobian");
+      petsc_object_state_increase(P);
+
+      // Handle the Jacobian-free case
+      // This call allows to resample the linearization point
+      // of the MFFD tangent operator
+      PetscBool flg;
+      AssertSNES(PetscObjectTypeCompare((PetscObject)A, MATMFFD, &flg));
+      if (flg)
+        {
+          AssertSNES(MatAssemblyBegin(A, MAT_FINAL_ASSEMBLY));
+          AssertSNES(MatAssemblyEnd(A, MAT_FINAL_ASSEMBLY));
+        }
+      else
+        petsc_object_state_increase(A);
+      PetscFunctionReturn(0);
+    };
+
+    auto snes_jacobian_with_setup_ =
+      [](SNES snes, Vec x, Mat A, Mat P, void *ctx) -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      (void)snes;
+      auto user = static_cast<NonlinearSolver *>(ctx);
+
+      VectorType  xdealii(x);
+      AMatrixType Adealii(A);
+      PMatrixType Pdealii(P);
+
+      user->A = &Adealii;
+      user->P = &Pdealii;
+      AssertUser(user->setup_jacobian(xdealii), "setup_jacobian");
+      petsc_object_state_increase(P);
+
+      // Handle older versions of PETSc for which we cannot pass a MATSHELL
+      // matrix to DMSetMatType
+      if (user->need_dummy_assemble)
+        {
+          AssertSNES(MatZeroEntries(P));
+          AssertSNES(MatAssemblyBegin(P, MAT_FINAL_ASSEMBLY));
+          AssertSNES(MatAssemblyEnd(P, MAT_FINAL_ASSEMBLY));
+          AssertSNES(MatShift(P, 0.0));
+        }
+
+      // Handle the Jacobian-free case
+      // This call allows to resample the linearization point
+      // of the MFFD tangent operator
+      PetscBool flg;
+      AssertSNES(PetscObjectTypeCompare((PetscObject)A, MATMFFD, &flg));
+      if (flg)
+        {
+          AssertSNES(MatAssemblyBegin(A, MAT_FINAL_ASSEMBLY));
+          AssertSNES(MatAssemblyEnd(A, MAT_FINAL_ASSEMBLY));
+        }
+      else
+        petsc_object_state_increase(A);
+      PetscFunctionReturn(0);
+    };
+
+    auto snes_monitor_ =
```

Comment:
For all of these local lambda functions, let's drop the trailing underscore.

Reply:
will do